File: Source/Fantasy.Base/XSerialization/XHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Reflection;
using System.ComponentModel;
using Fantasy.Properties;
using System.Xml.Linq;
using System.Xml;

namespace Fantasy.XSerialization
{
    public class XHelper
    {
        private  XHelper ()
	    {
            lock (_syncRoot)
            {
                foreach (Assembly asm in AppDomain.CurrentDomain.GetAssemblies())
                {
                    this.LoadXConverters(asm);
                }
                AppDomain.CurrentDomain.AssemblyLoad += new AssemblyLoadEventHandler(CurrentDomain_AssemblyLoad);
            }
	    }

        void CurrentDomain_AssemblyLoad(object sender, AssemblyLoadEventArgs args)
        {
            lock (_syncRoot)
            {
                this.LoadXConverters(args.LoadedAssembly);
            }
        }

        private void LoadXConverters(Assembly assembly)
        {
            var query = from type in assembly.GetTypes() where type.IsDefined(typeof(XConverterAttribute), false) select type;
            foreach (Type t in query)
            {
                XConverterAttribute attr = (XConverterAttribute)t.GetCustomAttributes(typeof(XConverterAttribute), false)[0];
                this._converters.Add(attr.TargetType, t);
            }
        }

        object _syncRoot = new object();
        

        private void LoadXPrefixes(Assembly assembly)
        {
            foreach(XPrefixAttribute attribute in assembly.GetCustomAttributes(typeof(XPrefixAttribute), true)) 
            {
                this._prefixes[attribute.Namesapce] = attribute.Prefix;
            }
        }

        private Dictionary<string, string> _prefixes = new Dictionary<string, string>();

        private Dictionary<Type, XTypeMap> _maps = new Dictionary<Type, XTypeMap>();
        private Dictionary<Type, Type> _converters = new Dictionary<Type, Type>();


        internal XTypeMap GetTypeMap(Type t)
        {
            lock (_syncRoot)
            {
                XTypeMap rs;
                if (this._maps.TryGetValue(t, out rs))
                {
                    return rs;

                }

                else if (t.IsDefined(typeof(XSerializableAttribute), false))
                {
                    rs = new XTypeMap(t);
                    this._maps.Add(t, rs);
                    return rs;
                }
                else
                {

                    throw new XException(String.Format(Resources.TypeIsNotXSerializableText, t));
                }
            }
        }


        public void LoadByXAttributes(IServiceProvider context, XElement element, object instance)
        {
            if (element == null)
            {
                throw new ArgumentNullException("element"); 
            }
            if (instance == null)
            {
                throw new ArgumentNullException("instance"); 
            }
            XTypeMap map = this.GetTypeMap(instance.GetType() );
            map.LoadByXAttributes(context, element, instance);
        }

        public void SaveByXAttributes(IServiceProvider context, XElement element, object instance)
        {
            if (element == null)
            {
                throw new ArgumentNullException("element");
            }
            if (instance == null)
            {
                throw new ArgumentNullException("instance");
            }
            XTypeMap map = this.GetTypeMap(instance.GetType());
            map.SaveByXAttributes(context, element, instance);
        }


        private bool IsNullableType(Type theType)
        {
            return (theType.IsGenericType && theType.GetGenericTypeDefinition().Equals(typeof(Nullable<>)));
        }

        private NullableConverter _nullableConvert = new NullableConverter();

        public TypeConverter CreateXConverter(Type t)
        {
            if (IsNullableType(t))
            {
                return _nullableConvert;
            }

            Type converterType = null;
            do
            {
                this._converters.TryGetValue(t, out converterType);
                t = t.BaseType;
            } while (converterType == null);

            return (TypeConverter)Activator.CreateInstance(converterType); 
        }

        private static XHelper _default = null;

        public static XHelper Default
        {
            get
            {
                if (_default == null)
                {
                    _default = new XHelper();
                }
                return _default;
            }
        }

        public static XmlNamespaceManager CreateXmlNamespaceManager(XElement element, bool includeAncestor)
        {
            XmlNamespaceManager rs = new XmlNamespaceManager(new NameTable());
            List<string> added = new List<string>();
            do
            {
                foreach (XAttribute attr in element.Attributes().Where(a => a.IsNamespaceDeclaration))
                {

                    string prefix = null;

                    if (attr.Name.LocalName == "xmlns" && attr.Name.NamespaceName == "")
                    {
                        prefix = "";

                    }
                    else if (attr.Name.Namespace == XNamespace.Xmlns)
                    {
                        prefix = attr.Name.LocalName;
                    }

                    if (prefix != null && added.IndexOf(prefix) < 0)
                    {
                        added.Add(prefix);
                        rs.AddNamespace(prefix, attr.Value);
                    }

                }

                element = element.Parent;
            } while (element != null && includeAncestor);
            return rs;
        }

        public static void AddNamespace(XmlNamespaceManager mngr, XElement element)
        {
            foreach (string prefix in mngr)
            {
                XNamespace ns = mngr.LookupNamespace(prefix);
                if (prefix == string.Empty)
                {
                    element.Add(new XAttribute("xmlns", ns));
                }

                else if (ns != XNamespace.Xmlns && ns != XNamespace.Xml)
                {
                    element.Add(new XAttribute(XNamespace.Xmlns + prefix, mngr.LookupNamespace(prefix)));
                }
            }
        }

    }
}

File: branches/Stheno/AutoConverted/Fantasy.Base/XSerialization/XHelper.java
package Fantasy.XSerialization;

import Fantasy.Properties.*;

public class XHelper
{
	private XHelper()
	{
		synchronized (_syncRoot)
		{
			for (Assembly asm : AppDomain.CurrentDomain.GetAssemblies())
			{
				this.LoadXConverters(asm);
			}
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
			AppDomain.CurrentDomain.AssemblyLoad += new AssemblyLoadEventHandler(CurrentDomain_AssemblyLoad);
		}
	}

	private void CurrentDomain_AssemblyLoad(Object sender, AssemblyLoadEventArgs args)
	{
		synchronized (_syncRoot)
		{
			this.LoadXConverters(args.LoadedAssembly);
		}
	}

	private void LoadXConverters(Assembly assembly)
	{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java:
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
		var query = from type in assembly.GetTypes() where type.IsDefined(XConverterAttribute.class, false) select type;
		for (java.lang.Class t : query)
		{
			XConverterAttribute attr = (XConverterAttribute)t.GetCustomAttributes(XConverterAttribute.class, false)[0];
			this._converters.put(attr.getTargetType(), t);
		}
	}

	private Object _syncRoot = new Object();


	private void LoadXPrefixes(Assembly assembly)
	{
		for(XPrefixAttribute attribute : assembly.GetCustomAttributes(XPrefixAttribute.class, true))
		{
			this._prefixes.put(attribute.getNamesapce(), attribute.getPrefix());
		}
	}

	private java.util.HashMap<String, String> _prefixes = new java.util.HashMap<String, String>();

	private java.util.HashMap<java.lang.Class, XTypeMap> _maps = new java.util.HashMap<java.lang.Class, XTypeMap>();
	private java.util.HashMap<java.lang.Class, java.lang.Class> _converters = new java.util.HashMap<java.lang.Class, java.lang.Class>();


	public final XTypeMap GetTypeMap(java.lang.Class t)
	{
		synchronized (_syncRoot)
		{
			XTypeMap rs = null;
			if ((rs = this._maps.get(t)) != null)
			{
				return rs;

			}

			else if (t.IsDefined(XSerializableAttribute.class, false))
			{
				rs = new XTypeMap(t);
				this._maps.put(t, rs);
				return rs;
			}
			else
			{

				throw new XException(String.format(Resources.getTypeIsNotXSerializableText(), t));
			}
		}
	}


	public final void LoadByXAttributes(IServiceProvider context, XElement element, Object instance)
	{
		if (element == null)
		{
			throw new ArgumentNullException("element");
		}
		if (instance == null)
		{
			throw new ArgumentNullException("instance");
		}
		XTypeMap map = this.GetTypeMap(instance.getClass());
		map.LoadByXAttributes(context, element, instance);
	}

	public final void SaveByXAttributes(IServiceProvider context, XElement element, Object instance)
	{
		if (element == null)
		{
			throw new ArgumentNullException("element");
		}
		if (instance == null)
		{
			throw new ArgumentNullException("instance");
		}
		XTypeMap map = this.GetTypeMap(instance.getClass());
		map.SaveByXAttributes(context, element, instance);
	}


	private boolean IsNullableType(java.lang.Class theType)
	{
		return (theType.IsGenericType && theType.GetGenericTypeDefinition().equals(.class));
	}

	private NullableConverter _nullableConvert = new NullableConverter();

	public final TypeConverter CreateXConverter(java.lang.Class t)
	{
		if (IsNullableType(t))
		{
			return _nullableConvert;
		}

		java.lang.Class converterType = null;
		do
		{
			converterType = this._converters.get(t);
			t = t.getSuperclass();
		} while (converterType == null);

		return (TypeConverter)Activator.CreateInstance(converterType);
	}

	private static XHelper _default = null;

	public static XHelper getDefault()
	{
		if (_default == null)
		{
			_default = new XHelper();
		}
		return _default;
	}

	public static XmlNamespaceManager CreateXmlNamespaceManager(XElement element, boolean includeAncestor)
	{
		XmlNamespaceManager rs = new XmlNamespaceManager(new NameTable());
		java.util.ArrayList<String> added = new java.util.ArrayList<String>();
		do
		{
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			for (XAttribute attr : element.Attributes().Where(a => a.IsNamespaceDeclaration))
			{

				String prefix = null;

				if (attr.getName().LocalName.equals("xmlns") && attr.getName().NamespaceName.equals(""))
				{
					prefix = "";

				}
				else if (attr.getName().Namespace == XNamespace.Xmlns)
				{
					prefix = attr.getName().LocalName;
				}

				if (prefix != null && added.indexOf(prefix) < 0)
				{
					added.add(prefix);
					rs.AddNamespace(prefix, attr.getValue());
				}

			}

			element = element.Parent;
		} while (element != null && includeAncestor);
		return rs;
	}

	public static void AddNamespace(XmlNamespaceManager mngr, XElement element)
	{
		for (String prefix : mngr)
		{
			XNamespace ns = mngr.LookupNamespace(prefix);
			if (prefix.equals(""))
			{
				element.Add(new XAttribute("xmlns", ns));
			}

			else if (ns != XNamespace.Xmlns && ns != XNamespace.Xml)
			{
				element.Add(new XAttribute(XNamespace.Xmlns + prefix, mngr.LookupNamespace(prefix)));
			}
		}
	}

}